core/src/layers/dtrace.rs

```diff
@@ -0,0 +1,227 @@
+/// Then you can use `readelf -n target/debug/examples/dtrace` to see the probes:
+///
+/// ```text
+/// Displaying notes found in: .note.stapsdt
+///   Owner                Data size        Description
+///   stapsdt              0x00000039       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: create_dir_start
+///     Location: 0x00000000000f8f05, Base: 0x0000000000000000, Semaphore: 0x00000000003649f8
+///     Arguments: -8@%rax
+///   stapsdt              0x00000037       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: create_dir_end
+///     Location: 0x00000000000f9284, Base: 0x0000000000000000, Semaphore: 0x00000000003649fa
+///     Arguments: -8@%rax
+///   stapsdt              0x0000003c       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: blocking_list_start
+///     Location: 0x00000000000f9487, Base: 0x0000000000000000, Semaphore: 0x0000000000364a28
+///     Arguments: -8@%rax
+///   stapsdt              0x0000003a       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: blocking_list_end
+///     Location: 0x00000000000f9546, Base: 0x0000000000000000, Semaphore: 0x0000000000364a2a
+///     Arguments: -8@%rax
+///   stapsdt              0x0000003c       NT_STAPSDT (SystemTap probe descriptors)
+/// ```
+#[derive(Default, Debug, Clone)]
+pub struct DTraceLayer {}
+
+impl<A: Accessor> Layer<A> for DTraceLayer {
+    type LayeredAccessor = DTraceAccessor<A>;
+    fn layer(&self, inner: A) -> Self::LayeredAccessor {
+        DTraceAccessor { inner }
+    }
+}
+
+#[derive(Clone)]
+pub struct DTraceAccessor<A: Accessor> {
+    inner: A,
+}
+
+impl<A: Accessor> Debug for DTraceAccessor<A> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DTraceAccessor")
+            .field("inner", &self.inner)
+            .finish_non_exhaustive()
+    }
+}
+
+#[async_trait]
+impl<A: Accessor> LayeredAccessor for DTraceAccessor<A> {
+    type Inner = A;
+    type Reader = A::Reader;
+    type BlockingReader = A::BlockingReader;
+    type Writer = A::Writer;
+    type BlockingWriter = A::BlockingWriter;
+    type Lister = A::Lister;
+    type BlockingLister = A::BlockingLister;
+    fn inner(&self) -> &Self::Inner {
+        &self.inner
+    }
+
+    async fn create_dir(&self, path: &str, args: OpCreateDir) -> Result<RpCreateDir> {
+        let c_path = CString::new(path).unwrap();
+        probe::probe_lazy!(opendal, create_dir_start, c_path.as_ptr());
+        let create_res = self.inner.create_dir(path, args).await;
+        probe::probe_lazy!(opendal, create_dir_end, c_path.as_ptr());
+        create_res
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let c_path = CString::new(path).unwrap();
+        probe::probe_lazy!(opendal, read_start, c_path.as_ptr());
+        let read_res = self.inner.read(path, args).await;
```

Comment:
It's better for use to provide `Reader`/`Writer`/`Lister` support. Please note that `read` just return a `Reader` to user, real IO doesn't happen here.